owncloud/src/main/java/ch/cyberduck/core/owncloud/OwncloudSession.java

```diff
@@ -57,23 +62,63 @@
 import ch.cyberduck.core.ssl.X509KeyManager;
 import ch.cyberduck.core.ssl.X509TrustManager;
 import ch.cyberduck.core.threading.CancelCallback;
+import ch.cyberduck.core.tus.TusCapabilities;
+import ch.cyberduck.core.tus.TusCapabilitiesResponseHandler;
+import ch.cyberduck.core.tus.TusWriteFeature;
 
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.client.HttpResponseException;
+import org.apache.http.client.methods.HttpOptions;
 import org.apache.http.impl.client.HttpClientBuilder;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import java.io.IOException;
+
 import com.auth0.jwt.JWT;
 import com.auth0.jwt.exceptions.JWTDecodeException;
 
+import static ch.cyberduck.core.tus.TusCapabilities.TUS_HEADER_RESUMABLE;
+import static ch.cyberduck.core.tus.TusCapabilities.TUS_VERSION;
+
 public class OwncloudSession extends DAVSession {
     private static final Logger log = LogManager.getLogger(OwncloudSession.class);
 
     private OAuth2RequestInterceptor authorizationService;
+    private HttpUploadFeature upload;
 
     public OwncloudSession(final Host host, final X509TrustManager trust, final X509KeyManager key) {
         super(host, trust, key);
     }
 
+    @Override
+    protected DAVClient connect(final Proxy proxy, final HostKeyCallback key, final LoginCallback prompt, final CancelCallback cancel) throws BackgroundException {
+        final DAVClient client = super.connect(proxy, key, prompt, cancel);
+        final TusCapabilities capabilities = this.options(client);
+        if(StringUtils.equals(TUS_VERSION, capabilities.version)) {
+            upload = new OcisUploadFeature(host, client.getClient(), new TusWriteFeature(capabilities, client.getClient()), capabilities);
+        }
+        else {
+            upload = new HttpUploadFeature(new NextcloudWriteFeature(this));
+        }
+        return client;
+    }
+
+    private TusCapabilities options(final DAVClient client) throws BackgroundException {
+        final HttpOptions options = new HttpOptions(URIEncoder.encode(
+                new DelegatingHomeFeature(new DefaultPathHomeFeature(host)).find().getAbsolute()));
+        options.setHeader(TUS_HEADER_RESUMABLE, TUS_VERSION);
```

Comment:
Not necessary for the `OPTIONS` request.
> The Tus-Resumable header MUST be included in every request and response except for OPTIONS requests

Reply:
I read that it in the sense that it *CAN* be included and thus we see the reason for the `OPTIONS` request.

Reply:
Don't get the point to add anything that is not needed and probably not evaluated on the servers side at this point.